validator/src/bootstrap.rs

```diff
@@ -1135,71 +1143,158 @@ mod with_incremental_snapshots {
 
-        let mut known_snapshot_hashes: HashMap<(Slot, Hash), HashSet<(Slot, Hash)>> =
-            HashMap::new();
-        known_validators
-            .iter()
-            .for_each(|known_validator| {
-                // First get the Crds::SnapshotHashes for each known validator and add them as
-                // the keys in the known snapshot hashes HashMap.
-                let mut full_snapshot_hashes = Vec::new();
-                cluster_info.get_snapshot_hash_for_node(known_validator, |snapshot_hashes| {
-                    full_snapshot_hashes = snapshot_hashes.clone();
-                });
-                full_snapshot_hashes.into_iter().for_each(|full_snapshot_hash| {
-                    // Do not add this hash if there's already a full snapshot hash with the same
-                    // slot but with a _different_ hash.
-                    // NOTE: There's no good reason for known validators to produce snapshots at
-                    // the same slot with different hashes, so this should not happen.
-                    if !known_snapshot_hashes.keys().any(|known_snapshot_hash| {
-                        known_snapshot_hash.0 == full_snapshot_hash.0 && known_snapshot_hash.1 != full_snapshot_hash.1
-                    }) {
-                        known_snapshot_hashes.insert(full_snapshot_hash, HashSet::new());
-                    } else {
-                        info!("Ignoring full snapshot hash from known validator {} with a slot we've already seen (slot: {}), but a different hash.", known_validator, full_snapshot_hash.0);
-                    }
-                });
+        'outer: for node in nodes {
+            // First get the full snapshot hashes for each node and add them as the keys in the
+            // golden snapshot hashes map.
+            let full_snapshot_hashes = get_full_snapshot_hashes_for_node(node);
+            for full_snapshot_hash in &full_snapshot_hashes {
+                // Do not add this hash if there's already a full snapshot hash with the same
+                // slot but with a _different_ hash.
+                // NOTE: Nodes should not produce snapshots at the same slot with _different_
+                // hashes.  So if it happens, keep the first and ignore the rest.
+                if !is_any_same_slot_and_different_hash(
+                    full_snapshot_hash,
+                    golden_snapshot_hashes.keys(),
+                ) {
+                    // Insert a new full snapshot hash into the golden snapshot hashes IFF an entry
+                    // doesn't already exist.  This is to ensure we don't overwrite existing
+                    // incremental snapshot hashes that may be present for this full snapshot hash.
+                    golden_snapshot_hashes
+                        .entry(*full_snapshot_hash)
+                        .or_default();
+                } else {
+                    debug!(
```

Comment:
I feel like this could be a `warn!()` actually, as it probably indicates one of your known validators is broken.  The node operator probably wants to know about this.

Reply:
Ok, will do.

Reply:
Done!